pkg/block/block.go

```diff
@@ -91,6 +93,20 @@ func Upload(ctx context.Context, logger log.Logger, bkt objstore.Bucket, bdir st
 		return errors.New("empty external labels are not allowed for Thanos block.")
 	}
 
+	if meta.Thanos.Files != nil {
+		return errors.New("meta.Thanos.Files section has to be empty; upload will handle this.")
+	}
+
+	meta.Thanos.Files, err = gatherFileStats(bdir)
+	if err != nil {
+		return errors.Wrap(err, "gather meta file stats")
+	}
+
+	metaEncoded := bytes.Buffer{}
+	if err := meta.Write(&metaEncoded); err != nil {
+		return errors.Wrap(err, "encode meta file")
+	}
+
 	if err := objstore.UploadFile(ctx, logger, bkt, path.Join(bdir, MetaFilename), path.Join(DebugMetas, fmt.Sprintf("%s.json", id))); err != nil {
```

Comment:
Here we upload meta.json file which isn't updated with files.

Reply:
Nice catch, moved metaEncoded just for this purpose but did not use it :scream: